Refactor Program Hooks (#161)
dustproject/dust: PROGRAMS.md

# WIP Programs: A Comprehensive Analysis

## Abstract

> This document provides a detailed examination of the Programs architecture within the DUST smart contract ecosystem. Programs represent a powerful abstraction layer that enables dynamic behavior and programmable logic for entities within the virtual world. By leveraging the MUD framework's capabilities, Programs offer a flexible mechanism for extending entity functionality through custom logic that can respond to various in-world events. This paper explores the theoretical foundations, implementation details, and practical applications of Programs, serving both as an academic analysis and a practical guide for developers.

Reviewer: I know this is not meant to be final but here are a few suggestions to make it sound a bit more down to earth haha
```suggestion
This document provides an overview of the Programs architecture within the DUST smart contract ecosystem. Programs represent a powerful abstraction layer that enables dynamic behavior and programmable logic for entities within the virtual world. By leveraging the MUD framework's capabilities, Programs offer a flexible mechanism for extending entity functionality through custom logic that can respond to various in-world events. This document explains the concepts, implementation details and practical applications of Programs, serving as a practical guide for developers.
```

Author: oh yeah this was mostly AI generated, just so we can at least have something, will include your changes

## 1. Introduction

In decentralized virtual worlds built on blockchain technology, the ability to create programmable objects with custom behaviors is essential for creating rich, interactive experiences. The Programs architecture in this project provides a framework for attaching custom logic to in-world entities, allowing them to respond to events, validate actions, and execute complex behaviors.

Programs function as an extension mechanism that separates an entity's data from its behavior, following principles similar to the Entity Component System (ECS) pattern. This separation enables dynamic composition of functionality and promotes code reuse across different entity types.

## 2. Framework

### 2.1 Core Concepts

Programs in this architecture are defined by several key abstractions:

> - **ProgramId**: A unique identifier that references a specific program implementation (the underlying value is a MUD's `ResourceId`)

Reviewer: ```suggestion
- **ProgramId**: A unique identifier that references a specific program implementation (the underlying value is a MUD `ResourceId`)
```

### 3.4 Gas Safety Mechanisms

To prevent malicious or poorly implemented programs from consuming excessive gas, calls with a fixed amount of gas are used. Safe calls are only used for hooks that correspond to actions that the entity might be incentivized to prevent (e.g. `onHit`).

The `SAFE_PROGRAM_GAS` constant (set to 1,000,000 gas units) limits the gas available to safe program calls, preventing the program from consuming all the gas available in the transaction.

## 4. Program Execution in Context

### 4.1 Machine Interactions

Programs can respond to interactions with machines, such as when a player fuels a machine:

```solidity
function fuelMachine(EntityId callerEntityId, EntityId machineEntityId, uint16 fuelAmount) public {
  // Validation and fuel logic
  // ...

  // Call program hook
  ProgramId program = baseEntityId.getProgram();
  program.callOrRevert(abi.encodeCall(IFuelHook.onFuel, (callerEntityId, baseEntityId, fuelAmount, "")));

  // Notification logic
  // ...
}
```

This allows machines to execute custom logic when they receive fuel, potentially triggering complex behaviors or state changes.

## 5. Practical Applications

> ### 5.1 Smart Objects

Reviewer: ```suggestion
### 5.1 Smart Entities
```

> Programs enable the creation of smart objects with custom behaviors:

Reviewer: ```suggestion
Programs enable the creation of smart entities with custom behaviors:
```

- **Security Systems**: Force fields that restrict actions based on complex rules
- **Interactive NPCs**: Non-player characters with programmable behavior

### 5.2 Gameplay Mechanics

Programs can implement complex gameplay mechanics:

- **Quest Systems**: Programs that track player progress and reward completion
- **Economic Systems**: Trading posts with dynamic pricing algorithms
- **Combat Systems**: Custom combat mechanics for specific areas or entities
- **Crafting Systems**: Advanced crafting pipelines that reward resource providers and periodically craft specific items

### 5.3 World Governance

Programs can facilitate governance mechanisms:

- **Access Control**: Programs that manage permissions for different areas
- **Resource Management**: Systems that regulate resource distribution
- **Voting Systems**: Mechanisms for collective decision-making
- **Taxation Systems**: Automated collection and distribution of resources

## 6. Implementation Guide

### 6.1 Creating a Program (WIP)

Example of a simple door program:

```solidity
contract SimpleDoorProgram is Program {
  constructor(IWorld world) Program(world) {}

>   function onAttach(EntityId caller, EntityId target, bytes memory extraData) external onlyWorld {

Reviewer: nit: was there a decision around renaming `extraData` to something more descriptive? (ie `programContext` or sth?)
```suggestion
  function onAttach(EntityId caller, EntityId target, bytes memory programContext) external onlyWorld {
```

Author: no decision but yeah I think we could do it, will leave for another pr so we can merge this soon

- **Resource Generators**: Machines that produce resources at regular intervals
- **Security Systems**: Force fields that restrict actions based on complex rules
- **Interactive NPCs**: Non-player characters with programmable behavior

### 5.2 Gameplay Mechanics

Programs can implement complex gameplay mechanics:

- **Quest Systems**: Programs that track player progress and reward completion
- **Economic Systems**: Trading posts with dynamic pricing algorithms
- **Combat Systems**: Custom combat mechanics for specific areas or entities
- **Crafting Systems**: Advanced crafting pipelines that reward resource providers and periodically craft specific items

### 5.3 World Governance

Programs can facilitate governance mechanisms:

- **Access Control**: Programs that manage permissions for different areas
- **Resource Management**: Systems that regulate resource distribution
- **Voting Systems**: Mechanisms for collective decision-making
- **Taxation Systems**: Automated collection and distribution of resources

## 6. Implementation Guide

### 6.1 Creating a Program (WIP)

Example of a simple door program:

```solidity

> contract SimpleDoorProgram is Program {

Reviewer 2: eg right now I have to figure out which hooks to add for a bed/force field. vs I could just do `is ForceFieldProgram`, and then I'd know the ones that I have to implement that will get called

> contract SimpleDoorProgram is Program {

Author: It only exists in our imagination

> contract SimpleDoorProgram is Program {

Author: I think we should eventually have a "dust standard library" for everything related to contracts that needs to be imported by users, we can include some utilities and a base Program contract with some helpers

> contract SimpleDoorProgram is Program {

Author: > eg right now I have to figure out which hooks to add for a bed/force field. vs I could just do is ForceFieldProgram, and then I'd know the ones that I have to implement that will get called

Sorry I missed this one, the thing is that all hooks should be optional right? like maybe you don't want to implement wakeup, just sleep. But yeah I guess with this design you could also provide base programs that group related hooks together, though I would still include the "default" fallback in them so that new hooks can be added in the future.

> contract SimpleDoorProgram is Program {

Reviewer 2: yee they're already optional cuz the physics doesnt require them. I mean in the "standard library" which is outside the physics it could have a base force field program contract, which is a guide to the user of which methods to implement. Otherwise as a user I have to figure that out myself
